National_Precinct_Bound_10_07_2022/main.ipynb

```diff
@@ -0,0 +1,266 @@
+    "\n",
+    "\n",
+    "def get_unique_col(precincts,election_cols):\n",
+    "    elec_copy = election_cols\n",
+    "    unique_id = False\n",
+    "    election_cols.append('geometry')\n",
+    "    for i in list(precincts.columns):\n",
+    "        if len(precincts[i].unique())==len(precincts):\n",
+    "            if i not in elec_copy:\n",
+    "                unique_id = i\n",
+    "                #print('Unique ID is: ' ,unique_id)\n",
+    "                break\n",
+    "    if unique_id == False:\n",
+    "        dict_of_lens ={}\n",
+    "        for i in precincts.columns:\n",
+    "            if i not in elec_copy:\n",
+    "                dict_of_lens.update({i:len(precincts[i].unique())})\n",
+    "        max1_len=0\n",
+    "        max2_len=0\n",
+    "        max1=False\n",
+    "        max2=False\n",
+    "        for k,v in dict_of_lens.items():\n",
+    "            if v>max1_len:\n",
+    "                max2_len = max1_len\n",
+    "                max2 = max1\n",
+    "                max1_len = v\n",
+    "                max1 = k\n",
+    "\n",
+    "            elif v>max2_len:\n",
+    "                max2_len = v\n",
+    "                max2 = k\n",
+    "            else:\n",
+    "                continue\n",
+    "        unique_id = str(max1)+'_'+str(max2)\n",
+    "        #print('Unique ID is: ', unique_id)\n",
+    "        precincts[unique_id] = precincts.apply(lambda x: str(x[max1])+ ' - '+str(x[max2]),axis=1)\n",
+    "    if len(precincts[unique_id].unique())!=len(precincts):\n",
+    "        precincts['counter'] = range(len(precincts))\n",
+    "        precincts[unique_id] = precincts.apply(lambda x: str(x[unique_id]) + ' - ' + str(x['counter']),axis=1)\n",
+    "    precincts['UNIQUE_ID'] = precincts[unique_id]\n",
+    "    #return unique_id\n",
+    "    #TODO: rename unique id column to name field \"UNIQUE_ID\" to return gdf, then concat on UNIQUE_ID\n",
+    "    return precincts\n",
+    "\n",
+    "\n",
+    "def get_election_cols(precincts):\n",
+    "    election_columns = []\n",
+    "    for i in list(precincts.columns):\n",
+    "        if i.startswith('G20'):\n",
+    "            election_columns.append(i)\n",
+    "    #print('Election columns: ', election_columns)\n",
+    "    return election_columns\n",
+    "\n",
+    "\n",
+    "def concat_precincts(state_abrvs):\n",
+    "    gdf_len = []\n",
+    "    gdf_concat = pd.DataFrame()\n",
+    "    for state in state_abrvs:\n",
+    "        gdf = gp.read_file(f'./prec_files/{state.lower()}_vest_20/{state.lower()}_vest_20.shp')\n",
+    "        gdf_len.append(gdf.shape[0])\n",
+    "        gdf['STATEABRV'] = state\n",
+    "        \n",
+    "        gdf = get_unique_col(gdf, get_election_cols(gdf))\n",
+    "        gdf['UNIQUE_ID'] = gdf['STATEABRV'] + '-' + gdf['UNIQUE_ID']\n",
+    "        gdf = gdf[['STATEABRV', 'UNIQUE_ID', 'geometry']].to_crs('EPSG:4326')\n",
```

Comment:
Seems good, if you did want to you could leave in the G20 Presidential fields since those should be uniform across the country.

Reply:
Kate said not necessary for this user RE election results.

Reply:
Good for now, let's add them in next week.